src/lib.rs

```diff
@@ -463,6 +468,122 @@ impl<'a, T: 'a> NodeMut<'a, T> {
         unsafe { self.tree.get_unchecked(self.id).has_children() }
     }
 
+    /// Apply function for each ancestor mutable node reference.
+    pub fn for_each_ancestor<F>(&mut self, mut f: F)
+    where
+        F: FnMut(&mut NodeMut<T>),
+    {
+        if let Some(mut parent_node) = self.parent() {
+            f(&mut parent_node);
+            while let Ok(mut n) = parent_node.into_parent() {
+                f(&mut n);
+                parent_node = n;
+            }
+        }
+    }
+
+    /// Apply function for each next sibling mutable node reference.
+    pub fn for_each_next_sibling<F>(&mut self, mut f: F)
+    where
+        F: FnMut(&mut NodeMut<T>),
```

Comment:
One other thing that applies to the other functions in the same manner might be that the lifetimes are more restrictive than necessary, e.g. wouldn't

```suggestion
    pub fn for_each_next_sibling<'a, F>(&'a mut self, mut f: F)
    where
        F: FnMut(&'a mut NodeMut<T>),
```

also work and enable references to escape the closure when that is sound?

Reply:
I've tried to do this, but hit borrow checker errors every time. It led to double mutable borrow and a short lifetime. Would love to be wrong, if somebody find how to solve lifetime constraints.

Reply:
Yeah, the problem is that we would need `into_*` variants of these traversal functions for that to work (and they would need to consume `self`). You should still be able to strengthen these signature to something like

```rust
pub fn for_each_ancestor<'b, F>(&'b mut self, mut f: F)
where
    F: FnMut(&mut NodeMut<'b, T>),
```

though.

Reply:
Did that.